README.md
# Permalinks Customizer

**Permalinks Customizer** helps you to set the different permalink structure for the PostTypes and Taxonomies which are publicly available. 

You can use the tags which are defined by the WordPress as mentioned [here](#default-tags-for-posttypes). Also, you can use the tags which are customily defined by this plugin for [PostTypes](#custom-tags-for-posttypes) and [Taxonomies](#custom-tags-for-taxonomies). You can also find these tags in Permalinks Customizer menu on WordPress Dashboard (if the plugin is installed).

## How to set the Permalinks for different PostTypes

Let's assume that you have 7 **PostTypes** and you like to apply different **permalink** structure for them. Like:

| PostType | Desired Permalink | Permalink Settings |
| ---------- | --------------------- | ---------------------- |
| Blog | http://www.example.com/blog/2019/01/02/this-is-a-first-post/ | blog/%year%/%monthnum%/%day%/%postname%/ |
| Customers | http://www.example.com/customers/postname/ | customers/%postname%/ |
| Events | http://www.example.com/events/2019-01-02-wordpress-meetup/ | events/%year%-%monthnum%-%day%-%postname%/ |
| Press | http://www.example.com/press/2019/announce-collaboration/ | press/%year%/%postname%/ |
| News | http://www.example.com/news/2019/this-is-a-news/ | news/%year%/%postname%/ |
| Sponsors | http://www.example.com/company/sponsor/yasglobal/ | company/%title%/ |
| Services | http://www.example.com/services/development/wordpress/ | services/%parent_postname%/ |

This plugin allows you to do this very easily. You just need to go on **Permalinks Customizer** Settings Page. Where text fields are shown with PostType name. You can define your permalinks you want to create for each post type.

If you leave any PostType field empty, **Permalinks Customizer** will create a permalink for that PostType by using the default **permalink** settings.

## How to Configure Permalinks Customizer
You can configure the plugin by navigating to the **Permalinks Customizer** menu from the WordPress Dashboard

## Structure Tags
You can find all the tags which are currently supported by the **Permalinks Customizer**.

### Default Tags for PostTypes
Below mentioned tags are provided by the **WordPress**. These tags can be used on Default **WordPress Permalink** Settings Page as well as **Permalinks Customizer** Settings page.

| Tag Name | Description |
| ----------- | ------------- |
| %year% | The year of the post, four digits, for example 2019 |
| %monthnum% | Month of the year, for example 01 |
| %day% | Day of the month, for example 02 |
| %hour% | Hour of the day, for example 15 |
| %minute% | Minute of the hour, for example 43 |
| %second% | Second of the minute, for example 33 |
| %post_id% | The unique ID of the post, for example 123 |
| %postname% | A sanitized version of the title of the post (post slug field on Edit Post/Page panel). So "This Is A Great Post!" becomes this-is-a-great-post in the URI. |
| %category% | A sanitized version of the category name (category slug field on New/Edit Category panel). Nested sub-categories appear as nested directories in the URI. |
| %author% | A sanitized version of the author name. |

### Custom Tags for PostTypes
Below mentioned tags are provided by the **Permalinks Customizer** for PostTypes. These tags can not be used on the Default **WordPress Permalink** Settings Page.

| Tag Name | Description |
| ----------- | ------------- |
| %title% | Title of the post. let's say the title is "This Is A Great Post!" so, it becomes this-is-a-great-post in the URI. |
| %parent_title% | This tag is similar as `%title%`.<br><br>Only the difference is that it appends the Immediate **Parent Page Title** in the URI if any parent page is selected. |
| %all_parents_title% | This tag is similar as `%title%`.<br><br>Only the difference is that it appends all the **Parents Page Title** in the URI if any parent page is selected. |
| %parent_postname% | This tag is similar as `%postname%`.<br><br>Only the difference is that it appends the Immediate **Parent Page Slug** in the URI if any parent page is selected. |
| %all_parents_postname% | This tag is similar as `%postname%`.<br><br>Only the difference is that it appends all the **Parents Page Slug** in the URI if any parent page is selected. |
| %child-category% | A sanitized version of the category name (category slug field on New/Edit Category panel). |
| %product_cat% | A sanitized version of the product category name (category slug field on New/Edit Category panel). Nested sub-categories appear as nested directories in the URI.<br><br>*This `tag` is specially used in WooCommerce Products.* |
| &lt;%ctax_custom_taxonomy%&gt; | A sanitized version of the custom taxonomy where the taxonomy name is `custom_taxonomy`. Replace the `custom_taxonomy` with your appropriate created taxonomy name.<br><br>If you want to provide the default slug which is used when the category/taxonomy doesn't be selected so, make sure to provide default name/slug which looks like this: `<%ctax_custom_taxonomy??sales%>`. Value which is written between the `??` and `%>` is used as default slug. |
| &lt;%ctaxparents_custom_taxonomy%&gt; | This tag is similar as `<%ctax_custom_taxonomy%>`.<br><br>Only the difference is that it appends all the **Parents Slug** in the URI if any parent category/tag is selected. |
| %author_firstname% | A sanitized version of the author first name. If author first name is not available then it uses the author's username. |
| %author_lastname% | A sanitized version of the author last name. If author last name is not available then it uses the author's username. |
| %pc_custom_posttype_tag% | Permits a theme or plugin developer define the tag value using a [filter](#set-custom-value-in-posttype-permalink) |

**Note**: `%title%` is similar as `%postname%` tag but the difference is that `%postname%` can only be set once by WordPress whereas `%title%` can be changed by user at multiple times. let's say the title is "This Is A Great Post!" so, it becomes "this-is-a-great-post" in the URI(At the first time, `%postname%` and `%title%` works same) but if you edit and change title let's say "This Is A WordPress Post!" so, `%postname%` in the URI remains same "this-is-a-great-post" whereas `%title%` in the URI becomes "this-is-a-wordpress-post".

### Custom Tags for Taxonomies
Below mentioned tags are provided by the **Permalinks Customizer** for Taxonomies. These can not be used on the Default **WordPress Permalink** Settings Page.

| Tag Name | Description |
| ----------- | ------------- |
| %name% | Name of the Term/Category. let's say the name is "External API" so, it becomes external-api in the URI. |
| %term_id% | The unique ID # of the Term/Category, for example 423 |
| %slug% | A sanitized version of the name of the Term/Category. So "External API" becomes external-api in the URI. |
| %parent_slug% | This tag is similar as `%slug%`.<br><br> Only the difference is that it appends the immediate **Parent Term/Category Slug** in the URI if any parent Term/Category is selected. |
| %all_parents_slug% | This tag is similar as `%slug%`.<br><br> Only the difference is that it appends all the **Parent Terms/Category Slugs** in the URI if any parent Term/Category is selected. |
| %pc_custom_taxonomy_tag% | Permits a theme or plugin developer define the tag value using a [filter](#set-custom-value-in-taxonomy-permalink) |

## Filters
You can find all the filters below which are provided by **Permalinks Customizer** plugin. These filters can be used as per your Website requirement.

### Set Custom value in PostType Permalink

This filter allow to replace the custom tag with your desired value. It can be any custom field value or anything else. 
```
/**
 * set the text which replace the custom tag from the permalink.
 *
 * @since 2.8.0
 *
 * @param object $post The post object.
 *
 * @return string text which can be replaced with the custom tag.
 */
function yasglobal_custom_posttype_tag( $post ) {
  return sanitize_title( $post->post_title ) . '-from-sami';
}
add_filter( 'pc_custom_posttype_tag', 'yasglobal_custom_posttype_tag', 10, 1 );
```

### Set Custom value in Taxonomy Permalink

This filter allow to replace the custom tag with your desired value. It can be any custom field value or anything else. 
```
/**
 * set the text which replace the custom tag from the permalink.
 *
 * @since 2.8.2
 *
 * @param object $term The term object.
 *
 * @return string text which can be replaced with the custom tag.
 */
function yasglobal_custom_taxonomy_tag( $term ) {
  return sanitize_title( $term->name ) . '-from-sami';
}
add_filter( 'pc_custom_taxonomy_tag', 'yasglobal_custom_taxonomy_tag', 10, 1 );
```

### Exclude Permalinks

If you want to exclude some Permalink to processed with the plugin so, just add the filter looks like this:
```
function yasglobal_exclude_url( $permalink ) {
  if ( false !== strpos( $permalink, '/contact-us/' ) ) {
    return '__true';
  }
  return;
}
add_filter( 'permalinks_customizer_exclude_request', 'yasglobal_exclude_url' );
```

### Show Relative Permalink/URL

To show relative permalink/url in Edit Post, add this filter in your themes functions.php.
```
add_filter( 'permalinks_customizer_remove_home_url', '__return_true' );
```

### Exclude PostType from the Plugin

To exclude the plugin to be worked on any PostType. Add this filter in your themes functions.php.

```
function yasglobal_exclude_post_types( $post_type ) {
  if ( $post_type == 'page' ) {
    return '__true';
  }
  return '__false';
}
add_filter( 'permalinks_customizer_exclude_post_type', 'yasglobal_exclude_post_types');
```
**Note**: Plugin stops working on the backend. *No more permalinks* would be generated by the plugin but the permalink which are already created will remains in work.

### Disable automatically create redirects

To disable automatically create redirects feature on creating and updating the post/pages/categories, add this filter in your themes functions.php.
```
add_filter( 'permalinks_customizer_auto_created_redirects', '__return_false');
```
This filter stops to be creating new redirects but existed redirects keeps working. To stop existed redirects, add [this](#disable-redirects) filter.

### Disable Redirects

To disable redirects to be applied , add this filter in your themes functions.php.
```
add_filter( 'permalinks_customizer_disable_redirects', '__return_false');
```
This filter only stop redirects to be work but the automatically create redirects still works. To stop automatically create redirects feature add [this](#disable-automatically-create-redirects) filter.

## Installation

This process defines you the steps to follow either you are installing through WordPress or Manually from FTP.

### From within WordPress

1. Visit 'Plugins > Add New'
2. Search for Permalinks Customizer
3. Activate Permalinks Customizer from your Plugins page.
4. Go to [after activation](#after-activation) below.

### Manually

1. Upload the `permalinks-customizer` folder to the `/wp-content/plugins/` directory
2. Activate Permalinks Customizer through the 'Plugins' menu in WordPress
3. Go to [after activation](#after-activation) below.

### After activation

1. Go to the plugin settings page and set up the plugin for your site.
2. You're done!

## Frequently Asked Questions

**Q. How to define Settings for the PostType?**

A. Navigate on **Permalinks Customizer** Menu from the **WordPress Dashboard**, Open PostTypes Settings Page, there is a textfield for each and every PostType (if the PostType is available for **Public**) . On this field, you can define structure which is used for that PostType.

**Q. Can i use tags in PostType Settings?**

A. Yes, you can use any tag which are defined in [Default Tags for PostTypes](#default-tags-for-posttypes) and [Custom Tags for Taxonomies](#custom-tags-for-taxonomies).

**Q. Does the plugin supports custom taxonomy tag?**

A. Yes, it supports the custom taxonomy tag as defined [here](#custom-tags-for-posttypes).

**Q. Can i see the created permalinks for the PostType?**

A. Yes, you can see all the created permalinks on the PostType Permalinks Page under Permalinks Customizer.

**Q. How to define Settings for the Taxonomies?**

A. Navigate on **Permalinks Customizer** Menu from the **WordPress Dashboard**, Open Taxonomies Settings Page, there is a textfield for each and every Taxonomy (if the Taxonomy is available for **Public**) . On this field, you can define structure which is used for that Taxonomy.

**Q. Can i use tags in Taxonomies Settings?**

A. Yes, you can use any tag which are defined [here](#custom-tags-for-taxonomies).

**Q. Can i see the created permalinks for the Taxonomies?**

A. Yes, you can see all the created permalinks on the Taxonomies Permalinks Page under Permalinks Customizer.

**Q. Can i regenerate all the permalinks according to the defined structure?**

A. Yes, you can regenerate all the permalinks according to the defined structure. To have a bulk permalink update, Go to the *All Post* page there is an option in the bulk action drop down with the name of `Regenerate Permalinks`. Use that option for regenerating the Permalinks.

**Q. Does *Regenerate Permalinks* damage my site SEO?**

A. No, it won't damage your site SEO. As regenerating permalinks added redirect against their previous permalink.

**Q. Can i see the available redirects?**

A. Yes, you can see the all the redirects created by this plugin from the Redirects Page under the Permalinks Customizer in the WordPress Dashboard.

**Q. Can i disable/delete redirects?**

A. Yes, you can disable/delete the redirects from the Redirects Page using Bulk Action.

**Q. Can i exclude PostType from the Plugin?**

A. Yes, you can exclude any posttype from the plugin to be worked on. For this just add the filter as shown [here](#exclude-posttype-from-the-plugin).


changelog.txt
== Changelog ==

This file contains only old changelog. See readme.txt for newer versions.

= 2.7.0 - Jul 05, 2019 =

* Enhancement
  * Added `%parent_title%` tag
  * Added `%all_parents_title%` tag
  * [Generate Permalink from code](https://wordpress.org/support/topic/how-to-regenerate-a-specific-post-id-in-my-function-php/)
  * Revamp `Tags` page
* Bug
  * [Flush Permalinks Cache](https://wordpress.org/support/topic/flush-permalinks-cache/)

= 2.6.0 - Jul 04, 2019 =

* Enhancement
  * Fix issue of Polylang `Hide Default` language Setting
  * Hide Default Permalink metabox when making changes in content
  * Compatible with `url_to_postid()`
* Bug
  * PHP Notice of [Undefined variable](https://github.com/yasglobal/permalinks-customizer/issues/26)
  * [Taxonomy permalink not working under existing page](https://wordpress.org/support/topic/taxonomy-permalink-not-working-under-existing-page/)

= 2.5.2 - Jan 02, 2019 =

* Enhancement
  * Added `<%ctaxparents_custom_taxonomy%>` tag for PostTypes
* Bug
  * Fixed PHP Fatal error in `uninstall.php`
    [https://wordpress.org/support/topic/unable-to-uninstall-10/](https://wordpress.org/support/topic/unable-to-uninstall-10/)

= 2.5.1 - Nov 30, 2018 =

* Enhancement
  * &lt;%ctax_custom_taxonomy%&gt; use category which is selected as Primary Category (Yoast SEO)

= 2.5.0 - Nov 28, 2018 =

* Enhancement
  * Auto refreshing permalink in Gutenberg
  * Removing need of Page refresh on `Regenerate Permalink` button
  * Added Permalink in Sidebar to Support Latest Gutenberg

= 2.4.0 - Oct 25, 2018 =

* Enhancement
  * Add `Flush Permalink Cache` option in Admin Toolbar
* Bug
  * `View Post` button does not open expected URL (due to relative Path)
    [https://wordpress.org/support/topic/view-post-button-does-not-open-expected-url/](https://wordpress.org/support/topic/view-post-button-does-not-open-expected-url/)
  * Removed auto flush rewrite rules to prevent permalink cache

= 2.3.1 - Oct 16, 2018 =

* Enhancement
  * Flush rewrite rules to prevent permalink cache
* Bug
  * Set meta_keys to be protected to stop duplication in Custom Fields

= 2.3.0 - Sept 10, 2018 =

* Enhancement
  * Added Support for Gutenberg
  * Added Privacy Policy Content on Admin Panel

= 2.2.0 - Aug 10, 2018 =

* Enhancement
  * Added Media (Attachment) Support
  * Added Regenerate Permalink in Bulk Action for Media (Attachment)
  * Added Filter to Exclude the PostType from the Plugin to be worked on
  * Added Support of `WP All Import` Plugin
  * Removing Permalink Edit Form from Private PostTypes and Taxonomies
* Bug
  * Plugin causing 504 gateway error when submitting a post for review
  * Prevent Permalink to be created for the private PostTypes like Coupon, Order etc

= 2.1.0 - July 16, 2018 =

* Enhancement
  * Generate Post Permalink as soon as Post saved at the very first time
  * Prevent to add auto-redirects for plain permalinks
  * Regenerate Permalink on Quick Edit Post
  * Added filter by which relative permalink can be shown on post edit page
* Bug
  * Regenerate Status Issue on Post/Page
  * Fixed Permalinks Customizer Version issue on plugin update
  * Removed Convert URL Page

= 2.0.1 - July 16, 2018 =

* Bugs
  * Fixed subdomain URL issue
  * Make the URLs to absolute on Admin Pages so, Permalinks for subdomain points to the correct URL

= 2.0.0 - July 09, 2018 =

* Enhancements
  * Added Custom Taxonomy Support for PostTypes
    * Define the Custom Taxonomy tag in PostType permalink structure
    * Define default taxonomy which is used in the permalink when no taxonomy is selected
  * Added `Regenerate Permalinks` in Bulk for PostTypes
  * Added `Regenerate Permalinks` in Bulk for Taxonomies
  * Added Redirect Functionality
    * Automatically create redirects on changing the permalink
    * Automatically create redirects on creating post/taxonomy
    * Automatically disable redirects if the newly created permalink was redireted before
    * Enable / Disable redirects from Plugin redirects Page
    * Delete redirects from Plugin redirects Page
    * Added Filters to disable redirects functionality if needed
  * Added Capabilities which allow users to view permalinks and manage settings without having administrator role
  * Reduce Query over non-admin pages
  * Avoid appending slashes and use trailingslashit instead
  * Removed unused PostTypes and Taxonomies (Only Public PostTypes and Taxonomies are shown on Settings Page)
* Bugs
  * Added a check for create term, to reduce undesired notifications.
  * Fixed undefined variable warning in PostType Permalinks and Taxonomies Permalinks Page
  * Fixed PHP warning on Bulk update over PostType Permalinks and Taxonomies Permalinks Page when no post/category is selected
  * Print `$view_post` varaible value

= 1.3.9 - Feb 13, 2018 =

* Enhancements
  * Fixed Empty Title issue
  * Added About Page
* Bugs
  * Fixed PHP Notices

= 1.3.8 - Oct 15, 2017 =

* Updated Postname if it's not available at the Created Post time

= 1.3.7 - Oct 06, 2017 =

* Resolved the Yoast SEO Slug Issue

= 1.3.6 - Oct 02, 2017 =

* Resolved the Draft Permalink Issue on editing the description under the snippet of Yoast SEO

= 1.3.5 - Sept 28, 2017 =

* Resolved the Pagination Issue for Taxonomies

= 1.3.4 - Sept 27, 2017 =

* Applied PHP Coding Standards on WordPress

= 1.3.3 - Sept 26, 2017 =

* Enhancements
  * Added Filter to Exclude/Ignore URL to be processed
* Bugs
  * Fixed Vulnerability Issues

= 1.3.2 - Sept 12, 2017 =

* Fixed Setting Custom Post Type Permalink doesn't work and redirects to /wp-admin/option

= 1.3.1 - Sept 11, 2017 =

* Added Translation Capability

= 1.3 - Sept 08, 2017 =

* Enhancements
  * Added PostTypes Permalinks Page
    * View all the PostTypes permalinks
    * Search Permalinks
    * Sort by Title
    * Pagination
  * Added Taxonomies Permalinks Page
    * View all the Category/Tags permalinks
    * Search Permalinks
    * Sort by Title
    * Pagination

* Bug Fixes

admin/class.permalinks-customizer-admin.php
<?php

/**
 * @package PermalinksCustomizer\Admin
 */

class Permalinks_Customizer_Admin {
  
	/**
	 * Initializes WordPress hooks
	 */
	function __construct() {
		add_action( 'admin_menu', array($this, 'permalinks_customizer_menu') );
		add_action( 'admin_init', array($this, 'register_permalinks_customizer_settings') );
  }

	/**
	 * Added Pages in Menu for Settings
	 */
	public function permalinks_customizer_menu() {
		add_menu_page( 'Set Your Permalinks', 'Permalinks Customizer', 'administrator', 'permalinks-customizer-posts-settings', array($this,'permalinks_customizer_posts_settings') );
		add_submenu_page( 'permalinks-customizer-posts-settings', 'Set PostTypes Permalinks', 'PostTypes Permalinks', 'administrator', 'permalinks-customizer-posts-settings', array($this, 'permalinks_customizer_posts_settings') );
		add_submenu_page( 'permalinks-customizer-posts-settings', 'Structure Tags for PostTypes', 'Tags for PostTypes', 'administrator', 'permalinks-customizer-post-tags', array( $this, 'permalinks_customizer_post_tags') );
		add_submenu_page( 'permalinks-customizer-posts-settings', 'Set Taxonomies Permalinks', 'Taxonomies Permalinks', 'administrator', 'permalinks-customizer-taxonomies-settings', array($this, 'permalinks_customizer_taxonomies_settings') );
		add_submenu_page( 'permalinks-customizer-posts-settings', 'Structure Tags for Taxonomies', 'Tags for Taxonomies', 'administrator', 'permalinks-customizer-taxonomy-tags', array( $this, 'permalinks_customizer_taxonomy_tags') );
		add_submenu_page( 'permalinks-customizer-posts-settings', 'Convert Custom Permalink', 'Convert Custom Permalink', 'administrator', 'permalinks-customizer-convert-url', array($this, 'permalinks_customizer_convert_url') );		
	}

	/**
	 * Register Fields to save settings of Post Types
	 */
	public function register_permalinks_customizer_settings() {
		$post_types = get_post_types( '', 'names' );
		foreach ( $post_types as $post_type ) {
			if ( $post_type == 'revision' || $post_type == 'nav_menu_item' || $post_type == 'attachment' ) {
				continue;
			}
			register_setting( 'permalinks-customizer-settings-group', 'permalinks_customizer_'.$post_type);
		}
	}

	/**
	 * Shows the main Settings Page Where user can provide different Permalink Structure for their Post Types
	 */
	public function permalinks_customizer_posts_settings() {   
		$post_types = get_post_types( '', 'objects' );
		echo '<div class="wrap">';
		echo '<h2>PostTypes Permalinks Settings</h2>';
		echo '<div>
						<p>Define the Permalinks for each post type. You can define different structures for each post type.</p>
						<p>Please check all the <a href="'.site_url().'/wp-admin/admin.php?page=permalinks-customizer-post-tags" title="structured tags">structured tags</a> which can be used with this plugin, <a href="'.site_url().'/wp-admin/admin.php?page=permalinks-customizer-post-tags" title="here">here</a>.</p>
					</div>';
		echo '<form method="post" action="options.php">';
		settings_fields( 'permalinks-customizer-settings-group' );
		do_settings_sections( 'permalinks-customizer-settings-group' );
		echo '<table class="form-table">';
		foreach ( $post_types as $post_type ) {
			if ( $post_type->name == 'revision' || $post_type->name == 'nav_menu_item' || $post_type->name == 'attachment' ) {
				continue;
			}
			$perm_struct = 'permalinks_customizer_'.$post_type->name;
			echo '<tr valign="top">
							<th scope="row">'.$post_type->labels->name.'</th>
							<td>'.site_url().'/<input type="text" name="'.$perm_struct.'" value="'.esc_attr( get_option($perm_struct) ) .'" class="regular-text" /></td>
						</tr>';
		}
		echo '</table>';
		submit_button(); 
		echo '</form>';
		echo '</div>';
	}

	/**
	 * Shows all the Tags which this Plugin Supports for Posts
	 */
	public function permalinks_customizer_post_tags() {
		$html  = '<div class="wrap">';
		$html .= '<h2>Structure Tags for Posts</h2>';
		$html .= '<div>These tags can be used to create Permalink Customizers for each post type.</div>';
		$html .= '<table class="form-table">';
		$html .= '<tr valign="top">
								<th scope="row">%title%</th>
								<td>Title of the post. let&#039;s say the title is "This Is A Great Post!" so, it becomes this-is-a-great-post in the URI.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%year%</th>
								<td>The year of the post, four digits, for example 2004</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%monthnum%</th>
								<td>Month of the year, for example 05</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%day%</th>
								<td>Day of the month, for example 28</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%hour%</th>
								<td>Hour of the day, for example 15</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%minute%</th>
								<td>Minute of the hour, for example 43</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%second%</th>
								<td>Second of the minute, for example 33</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%post_id%</th>
								<td>The unique ID # of the post, for example 423</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%postname%</th>
								<td>A sanitized version of the title of the post (post slug field on Edit Post/Page panel). So "This Is A Great Post!" becomes this-is-a-great-post in the URI.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%parent_postname%</th>
								<td>A sanitized version of the title of the post (post slug field on Edit Post/Page panel). So "This Is A Great Post!" becomes this-is-a-great-post in the URI. This <strong>Tag</strong> contains Immediate <strong>Parent Page Slug</strong> if any parent page is selected before publishing.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%all_parents_postname%</th>
								<td>A sanitized version of the title of the post (post slug field on Edit Post/Page panel). So "This Is A Great Post!" becomes this-is-a-great-post in the URI. This <strong>Tag</strong> contains all the <strong>Parent Page Slugs</strong> if any parent page is selected before publishing.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%category%</th>
								<td>A sanitized version of the category name (category slug field on New/Edit Category panel). Nested sub-categories appear as nested directories in the URI.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%child-category%</th>
								<td>A sanitized version of the category name (category slug field on New/Edit Category panel).</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%product_cat%</th>
								<td>A sanitized version of the product category name (category slug field on New/Edit Category panel). Nested sub-categories appear as nested directories in the URI. <i>This <strong>tag</strong> is specially used for WooCommerce Products.</i></td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%author%</th>
								<td>A sanitized version of the author name.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%author_firstname%</th>
								<td>A sanitized version of the author first name. If author first name is not available so, it uses the author\'s username.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%author_lastname%</th>
								<td>A sanitized version of the author last name. If author last name is not available so, it uses the author\'s username.</td>
							</tr>';
		$html .= '</table>';
		$html .= '<p><b>Note:</b> "%postname%" is similar as of the "%title%" tag but the difference is that "%postname%" can only be set once whereas "%title%" can be changed. let&#039;s say the title is "This Is A Great Post!" so, it becomes "this-is-a-great-post" in the URI(At the first time, "%postname%" and "%title%" works same) but if you edit and change title let&#039;s say "This Is A WordPress Post!" so, "%postname%" in the URI remains same "this-is-a-great-post" whereas "%title%" in the URI becomes "this-is-a-wordpress-post" </p>';
		$html .= '</div>';
		echo $html;
	}

	/**
	 * Shows the main Settings Page Where user can provide different Permalink Structure for their Taxonomies
	 */
	public function permalinks_customizer_taxonomies_settings() {   
		if (isset($_POST['submit'])) {
			$permalinks_customizer_taxes = array();
			foreach ($_POST as $key => $value) {
				if ($key === 'submit')
					continue;
				$permalinks_customizer_taxes[$key.'_settings'] = array(
					'structure' => $value,
				);
			}
			update_option('permalinks_customizer_taxonomy_settings', serialize( $permalinks_customizer_taxes ) );
		}
		$permalinks_customizer_settings = unserialize( get_option('permalinks_customizer_taxonomy_settings') );
		echo '<div class="wrap">';
		echo '<h2>Taxonomies Permalinks Settings</h2>';
		echo '<div>
						<p>Define the Permalinks for each taxonomy type so, the Permalinks would be created automatically on creating the Term/Category. Otherwise, you need to create the Permalinks manually from the Edit Term/Category Page.</p>
						<p>Please check all the <a href="'.site_url().'/wp-admin/admin.php?page=permalinks-customizer-taxonomy-tags" title="structured tags">structured tags</a> which can be used with this plugin, <a href="'.site_url().'/wp-admin/admin.php?page=permalinks-customizer-taxonomy-tags" title="here">here</a>.</p>
					</div>';
		echo '<form enctype="multipart/form-data" action="" method="POST">';
		echo '<table class="form-table">';
    $taxonomies = get_taxonomies(); 
		foreach ( $taxonomies as $taxonomy ) {
			if ( $taxonomy == 'nav_menu' ) {
				continue;
			}
			$value = '';
			if ( isset($permalinks_customizer_settings[$taxonomy.'_settings']) && isset($permalinks_customizer_settings[$taxonomy.'_settings']['structure']) && !empty($permalinks_customizer_settings[$taxonomy.'_settings']['structure']) ) {
				$value = $permalinks_customizer_settings[$taxonomy.'_settings']['structure'];
			}
			echo '<tr valign="top">
							<th scope="row">'.$taxonomy.'</th>
							<td>'.site_url().'/<input type="text" name="'.$taxonomy.'" value="'.$value.'" class="regular-text" /></td>
						</tr>';
		}
		echo '</table>';
		submit_button(); 
		echo '</form>';
		echo '</div>';
	}

	/**
	 * Shows all the Tags which this Plugin Supports for Taxonomies
	 */
	public function permalinks_customizer_taxonomy_tags() {
		$html  = '<div class="wrap">';
		$html .= '<h2>Structure Tags for Taxonomies</h2>';
		$html .= '<div>These tags can be used to create Permalink Customizers for each post type.</div>';
		$html .= '<table class="form-table">';
		$html .= '<tr valign="top">
								<th scope="row">%name%</th>
								<td>Name of the Term/Category. let&#039;s say the name is "External API" so, it becomes external-api in the URI.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%term_id%</th>
								<td>The unique ID # of the Term/Category, for example 423</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%slug%</th>
								<td>A sanitized version of the name of the Term/Category. So "External API" becomes external-api in the URI.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%parent_slug%</th>
								<td>A sanitized version of the name of the Term/Category. So "External API" becomes external-api in the URI. This <strong>Tag</strong> contains Immediate <strong>Parent Term/Category Slug</strong> if any parent Term/Category is selected before adding it.</td>
							</tr>';
		$html .= '<tr valign="top">
								<th scope="row">%all_parents_slug%</th>
								<td>A sanitized version of the name of the Term/Category. So "External API" becomes external-api in the URI. This <strong>Tag</strong> contains all the <strong>Parent Term/Category Slug</strong> if any parent Term/Category is selected before adding it.</td>
							</tr>';
		$html .= '</table>';
		$html .= '</div>';
		echo $html;
	}
	
	/**
	 * This Function Calls the another which provide the functionality to convert Custom Permalink URLs to Permalinks Customizer
	 */
	public function permalinks_customizer_convert_url() {
		require_once(PERMALINKS_CUSTOMIZER_PATH.'admin/class.permalinks-customizer-batch-script.php');
		new Permalinks_Customizer_Batch_Script();
	}
}
